File: Day14/convert-binary-number-in-a-linked-list-to-integer.cc
class Solution {
// Time: O(N) | Space: O(N)
public:
    int getDecimalValue(ListNode* head) {
        std::vector<int> values;
        while (head != nullptr) {
            values.push_back(head->val);
            head = head->next;
        }

        int currVal = 0;
        for (int val : values) {
            currVal = 2 * currVal + val;
        }
        return currVal;
    }
};

class Solution {
//  Time: O(N) | Space: O(1)
public:
    int getDecimalValue(ListNode* head) {
        int currVal = 0;
        while (head != nullptr) {
            currVal = 2 * currVal + head->val;
            head = head->next;
        }
        return currVal;
    }
};

File: Day14/convert-binary-number-in-a-linked-list-to-integer.js
function ListNode(val, next = null) {
    this.val = val;
    this.next = next;
}

//  Time: O(N) | Space: O(N)
var getDecimalValue = function(head) {
    const values = [];
    while (head !== null) {
        values.push(head.val);
        head = head.next;
    }

    let currVal = 0;
    for (let val of values) {
        currVal = 2 * currVal + val;
    }
    return currVal;
};

//  Time: O(N) | Space: O(1)
var getDecimalValue = function(head) {
    let currVal = 0;
    while (head !== null) {
        currVal = 2 * currVal + head.val;
        head = head.next;
    }
    return currVal;
};